spmd/tensor/parallel/api.py

```diff
@@ -137,51 +210,106 @@ def _parallelize_mlp(
             Tensor Parallelism on.
 
     Return:
-        None
+        A :class:``nn.Module`` object parallelized.
 
     .. warning::
         We only support ``PairwiseParallel`` right now.
     """
 
-    # Define partition functions needed.
-    def _rowwise_parallelize_fn(name, module, device_mesh):  # pyre-ignore[2, 3]
-        for name, param in module.named_parameters():
-            dist_spec = (
-                [Shard(1)] if name == "weight" else [Replicate()]  # type: ignore[list-item]
-            )
-            dist_param = torch.nn.Parameter(
-                distribute_tensor(param, device_mesh, dist_spec)
-            )
-            module.register_parameter(name, dist_param)
+    if not isinstance(parallel_style, PairwiseParallel):
+        raise NotImplementedError(
+            "Only support PairwiseParallel for Multihead Attention parallelization."
+        )
+
+    if device_mesh.ndim > 1:
+        device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
+
+    if isinstance(module, nn.MultiheadAttention):
+        tp_multi_head_attention = TensorParallelMultiheadAttention(
+            module.embed_dim,
+            module.num_heads,
+            device=torch.device(device_mesh.device_type),
+            tp_size=device_mesh.size(tp_mesh_dim),
+            add_bias_kv=module.bias_k is not None,
+        )
+        tp_multi_head_attention.copy(module)
+        return _parallelize_multihead_attn(tp_multi_head_attention, device_mesh)
```

Comment:
nit: instead of recursive call, shall we just assign module to `tp_multi_head_attention` instead?

Reply:
Oh I see your point, ok. Let me change it.